src/java/org/apache/cassandra/index/sai/disk/v3/CassandraDiskAnn.java

```diff
@@ -148,8 +160,11 @@ public CloseableIterator<ScoredRowId> search(float[] queryVector,
             }
             else
             {
-                scoreFunction = compressedVectors.approximateScoreFunctionFor(queryVector, similarityFunction);
-                reranker = i -> similarityFunction.compare(queryVector, view.getVector(i));
+                var sf = pqUnitVectors && similarityFunction == VectorSimilarityFunction.DOT_PRODUCT
+                         ? VectorSimilarityFunction.COSINE
+                         : similarityFunction;
+                scoreFunction = compressedVectors.approximateScoreFunctionFor(queryVector, sf);
+                reranker = i -> sf.compare(queryVector, view.getVector(i));
```

Comment:
Don't we want to use `similarityFunction` on the raw vector?

Reply:
you're right, thanks for catching that